src/optimizers/sr3.jl

```diff
@@ -1,52 +1,67 @@
-# Based upon alg 1 in
-# A unified sparse optimization framework to learn parsimonious physics-informed models from data
-# by K Champion et. al.
-
-mutable struct SR3{U,T} <: AbstractOptimizer
-    λ::U
-    ν::U
-    R::T
-end
-
+# Based upon alg 2 in https://ieeexplore.ieee.org/document/8573778
 
 """
-    SR3(λ, ν, R)
-    SR3(λ = 1e-1, ν = 1.0)
+$(TYPEDEF)
 
 `SR3` is an optimizer framework introduced [by Zheng et. al., 2018](https://ieeexplore.ieee.org/document/8573778) and used within
-[Champion et. al., 2019](https://arxiv.org/abs/1906.10612). `SR3` contains a sparsification parameter `λ`, a relaxation `ν`,
-and a corresponding penalty function `R`, which should be taken from [ProximalOperators.jl](https://github.com/kul-forbes/ProximalOperators.jl).
+[Champion et. al., 2019](https://arxiv.org/abs/1906.10612). `SR3` contains a sparsification parameter `λ`, a relaxation `ν`.
+
+It solves the following problem
+
+```math
+\\min_{x, w} \\frac{1}{2} \\| Ax-b\\|_2 + \\lambda R(w) + \\frac{\\nu}{2}\\|x-w\\|_2
+```
+
+Where `R` is a proximal operator and the result is given by `w`.
 
-# Examples
+#Fields
+$(FIELDS)
+
+# Example
 ```julia
 opt = SR3()
 opt = SR3(1e-2)
 opt = SR3(1e-3, 1.0)
+opt = SR3(1e-3, 1.0, SoftThreshold())
 ```
+
+## Note
+Opposed to the original formulation, we use `ν` as a relaxation parameter,
+as given in [Champion et. al., 2019](https://arxiv.org/abs/1906.10612). In the standard case of
+hard thresholding the sparsity is interpreted as `λ = threshold^2 / 2`, otherwise `λ = threshold`.
+
 """
-function SR3(λ = 1e-1, ν = 1.0)
-    R = NormL1
-    return SR3(λ, ν, R)
+mutable struct SR3{T,P} <: AbstractOptimizer where {T <: Real, P <: AbstractProximalOperator}
+    """Sparsity threshold"""
+    λ::T
+    """Relaxation parameter"""
+    ν::T
+    """Proximal operator"""
+    R::P
+
+    function SR3(threshold = 1e-1, ν = 1.0, R = HardThreshold())
+        λ = isa(R, HardThreshold) ? threshold^2 /2 : threshold
+        return new{typeof(λ), typeof(R)}(λ, ν, R)
+    end
 end
 
 function set_threshold!(opt::SR3, threshold)
-    opt.λ = threshold^2*opt.ν /2
+    opt.λ =  isa(opt.R, HardThreshold) ? threshold^2 /2 : threshold
     return
 end
 
-get_threshold(opt::SR3) = sqrt(2*opt.λ/opt.ν)
+get_threshold(opt::SR3) = opt.λ/opt.ν
 
 init(o::SR3, A::AbstractArray, Y::AbstractArray) =  A \ Y
 init!(X::AbstractArray, o::SR3, A::AbstractArray, Y::AbstractArray) =  ldiv!(X, qr(A, Val(true)), Y)
 
 function fit!(X::AbstractArray, A::AbstractArray, Y::AbstractArray, opt::SR3; maxiter::Int64 = 1, convergence_error::T = eps()) where T <: Real
-    f = opt.R(get_threshold(opt))
 
     n, m = size(A)
     W = copy(X)
 
     # Init matrices
-    P = inv(A'*A+I(m)/(opt.ν))
+    H = cholesky(A'*A+I(m)*opt.ν)
```

Comment:
this is guaranteed to be posdef?

Reply:
I've added assertions to make sure it is.